### ISSUE TITLE
Add PermitExtensions in scheduler to control when to admit the Pod binding process

### ISSUE THREAD

**Issue by kerthcet** [kubernetes/kubernetes#128903]
### What would you like to be added?

Let me introduce our user story first:

we have a model cache platform build on top of kubernetes. In model inference, when a Pod bind nodeName successfully and not started, we need to hang the pod from starting and syncing the model weights the inference pod needed from other nodes. To achieve this, we have to inject an initContainer to block pod starting and admit the initContainer until the model pulling finished.

This is **invasive** because we're modifying the Pod yaml and if the initContainer requires to watch CRDs, we may need to update the roles of the default SA for corresponding permissions, actually that's what we're faceing right now.

So we first think of an idea to block the sandbox creation by kubelet after Pod scheduled successfully, however, this may lead to pod resources reserved in scheduler but not consumed physically if pod hand forever for some reasons.

Then we came with another idea to extend the exist Permit Plugin in scheduler. So right now, permit plugin can return a `Wait` status in scheduling cycle, then in binding cycle, we'll block in `waitOnPermit` until condition ready. 

One example is coscheduling plugin, scheduling Pod could be blocked in bindingCycle until all its siblings are ready. 

However, based on my understanding, we have limitations here because once Pod is hanging in bindingCycle, we can only activate it in scheduling cycle by iterating over the waiting pods and return `Success` status. What if the Pod could be activated by other crd events rather than a new scheduling process?

So here comes a new proposal, we'll have a new interface `PermitExtensions`: 

```

type PermitFn func(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) error

type ClusterEventForPermission struct {
	Event    ClusterEvent
	PermitFn PermitFn
}

type PermitExtensions interface {
	Plugin
	RegisterPermissionEvents(context.Context) ([]ClusterEventForPermission, error)
}
```

The logic is quite similar to `EnqueueExtension`. First, we'll call the `RegisterPermissionEvents` to register events, and keep watching them in runtime. Once coming events match with the registered ones, the `PermitFn` will iterate over the waitingPods to decide whether Pods are ready or not, once ready, return `Success` status and the bindingCycle will be unblocked.

In our case, we have a CRD named `Torrent` who will record whether the model syncing task is done or not, and once completed, the Permit plugin's PermitFn will be triggered to decide whether we're ready to bind the Pod.

All existing flows will not be broken, no additional changes but offering a knob for users to decide when to admit a pod binding.

~~Then I came up with idea like maybe we can add some gates just like schedulingGates which will block the scheduling process, the new added gates (let's say xGates) will block the kubelet process, like creating the sandbox. Until the Pod holds a nodeName and no xGates exist, the kubelet will do his job.~~

~~The risk is since Pod is scheduled, the resource is reserved, but physically not consumed, so if the postScheduling process blocks, resource is wasted, what we have to do next is recreate the Pod (pod.nodeName is unchangeable now if set, maybe we can unbind the Pod and don't change the nodeName anymore until `nodeName != "" && xGates == []`, this is a big change I think).~~

~~Another approach would be introduce another logic process in scheduler, and hold the bind process until the condition is ready, quite like the DRA, in DRA, we have a lot of interactions between the scheduler and the dra controller, until everything is ready, open fire. DRA is a huge work, I hope not to introduce more CRDs, so maybe we can add new gates like bindingGates here? Extending the schedulingGates is ambiguous I think.~~

Some rough thoughts here, would like to hear advices or suggestions. Thanks in advance!

cc @liggitt @SergeyKanzhelev @kubernetes/sig-scheduling-misc 
/kind feature
/sig node
/sig scheduling

### Why is this needed?

Once Pod is assumed to be bind with a node, or already bind with a node, we may have more work to do before pod starting, here we introduce a mechanism to make this happen before pod binding.

**Comment by sanposhiho** [comment]
> To achieve this, we have to inject an initContainer to user Pods and admit the Pod until the model pulling finished.

So, who currently pulls models? initContainer?

**Comment by haosdent** [comment]
I am not sure if you could make your model become a docker image, and them mount it as volume via this way https://kubernetes.io/docs/tasks/configure-pod-container/image-volumes/
Then, we could avoid modifying the user's YAML file.

**Comment by haosdent** [comment]
> make your model become a docker image

Another reason is that we could leverage existing Image P2P distribution solutions, such as Dragonfly, to speed up the model distribution.

**Comment by pohly** [comment]
> in DRA, we have a lot of interactions between the scheduler and the dra controller, until everything is ready, open fire

Not anymore. DRA controller was part of classic DRA, which got removed in 1.32.

Now DRA only needs to do some additional writes in PreBind to update ResourceClaim status.

**Comment by Huang-Wei** [comment]
I'm not quite sure I fully understand the problem statement - are you trying to say before the pod starts to run on the target node, you'd like a way to download the model on the target node? and seems you'd assume the model is sharable on the node?

**Comment by pacoxu** [comment]
Perhaps we don't have to limit ourselves to just the scenario of large model file download preparation.

Other possible scenarios:

- PodGroup could also be a scenario where a group of Pods are scheduled and started simultaneously, and some Pods may need to wait for others to be successfully scheduled.

I'm not sure if there are any other similar scenarios.

**Comment by sanposhiho** [comment]
I don't understand what this issue is trying to say yet. PodGroup stuff is different since it's already implementable with Permit extension point, like coscheduling plugin shows. 
(... and @kerthcet must know it, so I suppose he's trying to mean something else here.)

**Comment by kannon92** [comment]
We ran into this problem for KubeFlow Training. https://github.com/kubernetes-sigs/jobset/issues/672

For training we have a replicated job that runs first and we wait for that job to finish before proceeding.

Are you doing this also for inference?

**Comment by alculquicondor** [comment]
This sounds more like
/sig node
/wg serving

> The risk is since Pod is scheduled, the resource is reserved, but physically not consumed

So what is the proposal? What would the system do differently if there's an xGate in the Pod? Should we reserve the resources or not?

**Comment by kerthcet** [comment]
Sorry, I'm OOO, let me explain more details.

So what we **want** is when a Pod is bind successfully with a node, **before** the kubelet start to create the sandbox, an additional controller will be responsible for downloading/syncing model weights from other nodes. At that point, we don't want the Pod to start, because obviously the Pod _wouldn't_ start successfully. So if we can have a hook here as I mentioned, like a xGate, this signal will tell kubelet when to start or stop.
- If xGate exists, the kubelet stop creating the sandbox
- With xGates removed, kubelet stops to create the sandbox.

That's what I mean post scheduling.

**Comment by kerthcet** [comment]
So generally, what we do right now is **Prepare everything and schedule the Pod to the fit node**, maybe there're some other cases like **Schedule the Pod to the constraint-fit node, and prepare the other prerequisites before starting the Pod**? I don't know if others have similar cases, just put this here for discussion.

**Comment by kerthcet** [comment]
I put the sig scheduling here is because it's possible to implement this in the scheduler as well, like we have bindingGates, and expose the candidate node info as well, the default bind api will block here if we have bindingGates, the external controller will watch for the candidate node info and prepare everything before removing the bindingGates, in our case, the candidate node will try to syncing model weights from other node peers.

Once bindingGate is removed, we'll admit the bind api and everything goes normal.

**Comment by sanposhiho** [comment]
It somehow reminded me of the past reservation proposal.
https://docs.google.com/document/d/1sbFUA_9qWtorJkcukNULr12FKX6lMvISiINxAURHNFo/edit?tab=t.0#heading=h.xgjl2srtytjt

With it, we can do something like
1. You create `Reservation` and it is scheduled on a Node.
2. The controller checks where `Reservation` goes and does the model pulling stuff.
3. Once finishes, the controller creates a Pod to consume the reservation.

I know `xGate` is much easier to implement though. Just to present an alternative option.

**Comment by sanposhiho** [comment]
.. and similarly, you can do a trick with ballon Pods to solve your problem, even today:

1. You create a ballon Pod and it's scheduled on a Node.
2. The controller checks where the ballon pod goes and does the model pulling stuff.
3. Once finishes, the controller creates a real Pod with a node affinity to the Node / with a high priority to take the place of a ballon Pod.

**Comment by alculquicondor** [comment]
I suggest you bring this up to the WG Serving for brainstorming.

Could you use an initContainer that uses fsNotify to wait for the model to be available in the filesystem? This wouldn't require any permissions.

> the default bind api will block here if we have bindingGates

I don't think this can be considered a scheduling feature. Scheduling is concerned about assigning a pod to a node, not pod startup.

**Comment by kerthcet** [comment]
> It somehow reminded me of the past Pod Resource Reservation discussion.

If the PRR is considered by the kubelet as well, we may make it out, but as I recalled, there's no abstraction about Pod so the change is huge and not accepted. We played a workaround to make this happen but hope to make it extensible, but still thanks for the suggestions.

**Comment by kerthcet** [comment]
> I don't think this can be considered a scheduling feature. Scheduling is concerned about assigning a pod to a node, not pod startup.

We can make this happen before binding, like extending the Permit plugin, right now, the waiting Pods can only be admitted by its siblings in scheduling cycle (I didn't see other implementations of Permit plugin), maybe we can extend the capacity to **admit the waiting pods by specified events**, which acts like bindingGates I mentioned. I haven't thought out the whole implementation details yet, just a bold idea.

Then I can write a plugin to achieve what I expected, although still a little trouble for our users because they have to use the customized scheduler, but at least a feasible way.

**Comment by kerthcet** [comment]
> Could you use an initContainer that uses fsNotify to wait for the model to be available in the filesystem? This wouldn't require any permissions.

The simple answer is **better not** just because the initContainer has no idea how much files(model weights here) it needs unless it sends a query.

**Comment by kerthcet** [comment]
They're just two different approaches, I'm update it once I have a plan.

**Comment by kerthcet** [comment]
Here's the latest proposal for scheduling feature update, copied from the updated issue descriptions:


We came with another idea to extend the exist Permit Plugin in scheduler. So right now, permit plugin can return a Wait status in scheduling cycle, then in binding cycle, we'll block in `waitOnPermit` until condition ready.

One example is coscheduling plugin, scheduling Pod could be blocked in bindingCycle until all its siblings are ready.

However, based on my understanding, we have limitations here because once Pod is hanging in bindingCycle, we can only activate it in scheduling cycle by iterating over the waiting pods and return Success status. What if the Pod could be activated by other crd events rather than a new scheduling process?

So here comes a new proposal, we'll have a new interface `PermitExtensions`:

```
type PermitFn func(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) error

type ClusterEventForPermission struct {
	Event    ClusterEvent
	PermitFn PermitFn
}

type PermitExtensions interface {
	Plugin
	RegisterPermissionEvents(context.Context) ([]ClusterEventForPermission, error)
}
```

The logic is quite similar to EnqueueExtension. First, we'll call the RegisterPermissionEvents to register events, and keep watching them in runtime. Once coming events match with the registered ones, the PermitFn will iterate over the waitingPods to decide whether Pods are ready or not, once ready, return Success status and the bindingCycle will be unblocked.

In our case, we have a CRD named `Torrent` who will record whether the model syncing task is done or not, and once completed, the Permit plugin's PermitFn will be triggered to decide whether we're ready to bind the Pod.

All existing flows will not be broken, no additional changes for current user but offering a knob for users to decide when to admit a pod binding.

What do you guys think?

**Comment by sanposhiho** [comment]
Why not just make the plugin watch the CRD and activate waiting pods as necessary, when it's ready?

**Comment by kerthcet** [comment]
Do you mean have a background task watching the CRD? The scheduling cycle is non-blocking and we didn't provide ways to do that asynchronous, that's the proposal hope to offer.

**Comment by sanposhiho** [comment]
> Do you mean have a background task watching the CRD? 

Yes, the plugin starts EventHandler or whatever at its `New` func.

**Comment by kerthcet** [comment]
I don't think we're happy to see plugins have there own background tasks, it's heavy and risky, as a plugin system, I believe it's better to be stateless. And it's our duty to abstract the framework to satisfy different scenarios, our case is, maybe others as well, that's why I asked feedbacks here, I will ask in the serving WG as well but honestly this is not only related to serving.

**Comment by alculquicondor** [comment]
> They're just two different approaches to solve the same problem

Perhaps. But I'm not talking about **how** we implement the feature. I'm talking about **what** the feature does.
The feature is related to pod lifecycle, thus, it should be consulted with sig node.

**Comment by alculquicondor** [comment]
That said, I believe the scheduler is not the right place to put such a feature.

**Comment by kerthcet** [comment]
Agree, sig node should be part of the discussion, cc @kubernetes/sig-node-feature-requests for more inputs. I think @kannon92 and @pacoxu are already involved from sig node side. And I also mentioned this in serving WG.

But I disagree only sig node should be responsible for pod lifecycle, sig node cares more about the interactions between pod and host, part of pod lifecycle, but scheduling is also part of that. Particularly, in my new proposal, I'm talking about adding some hooks in Permit extension point to make it possible for plugins to control the admission time point (unblocking the waiting pods). From both the domain and code perspectives, I think sig scheduling should also be involved. And actually we're, that's why we're discussing here, appreciate all the opinions.

I'm still hoping to see more use cases.

**Comment by alculquicondor** [comment]
> I'm talking about adding some hooks in Permit extension point to make it possible for plugins to control the admission time point
The question would be what for. Your use case is specifically about pulling a model. You could maybe extend this to pulling an image. Do you have other use cases that justify this is a Permit extension?

**Comment by kerthcet** [comment]
Pity to say not yet, put the discussion here for more inputs.
